src/dmd/root/file.d

```diff
@@ -22,6 +22,379 @@ import dmd.root.filename;
 import dmd.root.rmem;
 import dmd.root.string;
 
+/**
+Encapsulated management of a memory-mapped file.
+
+The `Datum` type encodes the mapping mode: Use `ubyte` for read/write mapping
+and `const ubyte` for read-only mapping.
+*/
+struct FileMapping(Datum)
```

Comment:
The `ubyte -> RW` / `const ubyte -> R` mapping is nice, however, someone that sees `FileMapping!(ubyte)("bla.txt")` might not understand it right away. `FileMapping!(Read)`/`FileMaping(ReadWrite)` reads better in my opinion. Also, what happened to plain `Write` ?

Reply:
The idea is to allow support for other units such as `uint` or `ulong`. Also, most of the code is common to the read and read-write modes so a template was scoring on both these points.

There's no write-only mapping as far as I understand.

Reply:
> The idea is to allow support for other units such as uint or ulong.

Martin says don't build in features 'till ya need them. You're going to be sorry you gave me that book :-)

It makes the interface "heavier", and the unused features likely don't even work because they are never tested. It will be trivial to add this if/when it is needed.

Reply:
> Martin says don't build in features 'till ya need them.

Actually that's exactly what the code is doing: it rejects data types that are not currently used. The template is necessary because there are two modes that have 90% identical code and 99% identical interface.

> It makes the interface "heavier"

In fact the opposite is true. The interface is virtually identical across the read-only and read-write versions (only the `resize` function is not offered in read-only mode). Semantics are identical, too. So the reader only needs to read and absorb ONE interface to understand TWO ways of using memory-mapped files, as opposed to looking at TWO distinct interfaces (subtle duplication for the lose!) that it just so happens are 99% identical with no indication of that in the source code.

Template is the way to go here, no two ways about it. I could use a `bool` or `enum` to say "read-only vs read-write" but that is the same aggravation for less upside.